### PR TITLE
loader: fix loader painc when not remove meta

### LINKED ISSUE #1770: loader panic
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

loader painc:
`[2021/06/15 11:27:33.117 +08:00] [ERROR] [db.go:177] ["execute statements failed after retry"] [task=test-all] [unit=load] ["worker ID"=10] [queries="[INSERT INTO `dm_meta`.`test-all_loader_checkpoint` (`id`, `filename`, `cp_schema`, `cp_table`, `offset`, `end_pos`) VALUES(?,?,?,?,?,?)]"] [arguments="[[mysql-replica-01 dm_benchmark.sbtest4.0000000020015.sql dm_benchmark sbtest4 0 1048711]]"] [error="[code=10006:class=database:scope=not-set:level=high], Message: execute statement failed: commit, RawCause: Error 1062: Duplicate entry 'mysql-replica-01-dm_benchmark.sbtest4.0000000020015.sql' for key 'uk_id_f'"]
[2021/06/15 11:27:33.119 +08:00] [ERROR] [checkpoint.go:354] ["checkpoint record already exists, skip it."] [task=test-all] [unit=load] [component="remote checkpoint"] [id=mysql-replica-01] [filename=dm_benchmark.sbtest4.0000000020015.sql]
panic: runtime error: index out of range [0] with length 0`

```
goroutine 7918 [running]:
github.com/pingcap/dm/loader.(*RemoteCheckPoint).UpdateOffset(0xc000556200, 0xc001574b95, 0x26, 0xf42f5)
        /Users/ehco/Documents/pingcap/dm/loader/checkpoint.go:406 +0x376
github.com/pingcap/dm/loader.(*Worker).run.func2()
        /Users/ehco/Documents/pingcap/dm/loader/loader.go:202 +0x410
github.com/pingcap/dm/loader.(*Worker).run.func3(0xc0014cae40, 0xc001622020)
        /Users/ehco/Documents/pingcap/dm/loader/loader.go:222 +0x55
created by github.com/pingcap/dm/loader.(*Worker).run
        /Users/ehco/Documents/pingcap/dm/loader/loader.go:220 +0x27e
```

4. Versions of the cluster

    - DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):

        ```
        Release Version: v2.0.0-beta.2-426-g26e4d932
        Git Commit Hash: 26e4d9325935acd61388d2dc31fa2b1dbf9a3047
        Git Branch: lag
        UTC Build Time: 2021-06-09 09:25:27
        Go Version: go version go1.16.3 darwin/amd64
        ```

    - Upstream MySQL/MariaDB server version:

        ```
        5.7.34
        ```

    - Downstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

        ```
        5.0.2
        ```

    - How did you deploy DM: DM-Ansible or manually?

        ```
        manually
        ```

5. current status of DM cluster (execute `query-status` in dmctl)
```
{
    "result": true,
    "msg": "",
    "sources": [
        {
            "result": false,
            "msg": "[code=38029:class=dm-master:scope=internal:level=high], Message: mysql-replica-01 relevant worker-client not found",
            "sourceStatus": {
                "source": "mysql-replica-01",
                "worker": "",
                "result": null,
                "relayStatus": null
            },
            "subTaskStatus": [
            ]
        }
    ]
}
```

6. Operation logs
   - Please upload `dm-worker.log` for every DM-worker instance if possible
   - Please upload `dm-master.log` if possible
   - Other interesting logs
   - Output of dmctl's commands with problems
   
7. Configuration of the cluster and the task
   - `dm-worker.toml` for every DM-worker instance if possible
   - `dm-master.toml` for DM-master if possible
   - task config, like `task.yaml` if possible
   - `inventory.ini` if deployed by DM-Ansible

8. Screenshot/exported-PDF of Grafana dashboard or metrics' graph in Prometheus for DM if possible

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve? 

close #1770

### What is changed and how it works?

when loader checkpoint `UpdateOffset` meet error, just return that error and Pause the task

### Check List 

Tests 

 - Integration test

Code changes

 - Has exported function/method change
 - Has interface methods change

Side effects
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: tests/load_interrupt/run.sh [hunk 1]
```diff
@@ -33,12 +33,43 @@ function check_row_count() {
...
 }
 
-function run() {
-	THRESHOLD=1024
+function test_save_checkpoint_faild() {
 	prepare_datafile
-
 	run_sql_file $WORK_DIR/db1.prepare.sql $MYSQL_HOST1 $MYSQL_PORT1 $MYSQL_PASSWORD1
 
+	export GO_FAILPOINTS="github.com/pingcap/dm/loader/loaderCPUpdateOffsetError=return(1)"
```
Comment (lance6716) [inline tests/load_interrupt/run.sh:40]: I remember failpoint could use `return()` to trigger an inject, not sure if there's a more simple way

**Review by lance6716 — APPROVED** [review]

File: loader/checkpoint.go [hunk 2]
```diff
@@ -395,15 +395,23 @@ func (cp *RemoteCheckPoint) GenSQL(filename string, offset int64) string {
...
 // UpdateOffset implements CheckPoint.UpdateOffset.
-func (cp *RemoteCheckPoint) UpdateOffset(filename string, offset int64) {
+func (cp *RemoteCheckPoint) UpdateOffset(filename string, offset int64) error {
 	cp.restoringFiles.Lock()
 	defer cp.restoringFiles.Unlock()
 	db, table, err := getDBAndTableFromFilename(filename)
 	if err != nil {
-		cp.logger.Error("error in checkpoint UpdateOffset", zap.Error(err))
-		return
+		return terror.ErrLoadTaskCheckPointNotMatch.Generate()
```
Comment (GMHDBJD) [inline loader/checkpoint.go:403]: It's different error here.

File: errors.toml [hunk 3]
```diff
@@ -1528,6 +1528,12 @@ description = ""
 workaround = "Please check if the previous worker is online."
 tags = ["internal", "high"]
 
+[error.DM-functional-34018]
+message = "inconsistent checkpoints between loader and target database"
+description = ""
+workaround = "Please check that you have not forgotten to use dmctl start-task --remove-meta=true."
```
Comment (GMHDBJD) [inline errors.toml:1534]: Maybe we should add some prerequisites. e.g. If you want to redo the task.

File: loader/checkpoint.go (hunk 2 above)
Comment (Ehco1996) [reply loader/checkpoint.go:403]: i think's it's a same error that  is an error caused by an incorrect filename？

Comment (GMHDBJD) [reply loader/checkpoint.go:403]: How about use terror.Annotate?

Comment (Ehco1996) [reply loader/checkpoint.go:403]: you mean annotate two error with different  error msg ?

File: errors.toml [hunk 4]
```diff
@@ -1528,6 +1528,12 @@ description = ""
 workaround = "Please check if the previous worker is online."
 tags = ["internal", "high"]
 
+[error.DM-functional-34018]
+message = "inconsistent checkpoints between loader and target database"
+description = ""
+workaround = "Please check that you have not forgotten to use dmctl start-task --remove-meta=true. if you want to redo the task."
```
Comment (GMHDBJD) [inline errors.toml:1534]: If you want to redo the whole task, please check that you have not forgotten to add `-remove-meta` flag for `start-task` command.

File: loader/loader.go [hunk 5]
```diff
@@ -212,7 +212,17 @@ func (w *Worker) run(ctx context.Context, fileJobQueue chan *fileJob, runFatalCh
 				})
 				continue
 			}
-			w.loader.checkPoint.UpdateOffset(job.file, job.offset)
+
+			failpoint.Inject("loaderCPUpdateOffsetError", func(_ failpoint.Value) {
+				job.file = "notafile" + job.file
+			})
+			if err := w.loader.checkPoint.UpdateOffset(job.file, job.offset); err != nil {
+				if !utils.IsContextCanceledError(err) {
```
Comment (GMHDBJD) [inline loader/loader.go:220]: Seems no need check context cancel.

**Review by GMHDBJD — COMMENTED** [review]
rest LGTM

File: loader/checkpoint.go [hunk 6]
```diff
@@ -395,15 +395,23 @@ func (cp *RemoteCheckPoint) GenSQL(filename string, offset int64) string {
...
+
+	if _, ok := cp.restoringFiles.pos[db]; ok {
+		if _, ok := cp.restoringFiles.pos[db][table]; ok {
+			if _, ok := cp.restoringFiles.pos[db][table][filename]; ok {
+				cp.restoringFiles.pos[db][table][filename][0] = offset
+				return nil
+			}
+		}
+	}
+	return terror.Annotatef(terror.ErrLoadTaskCheckPointNotMatch.Generate(), "db=%s table=%s not in checkpoint", db, filename)
```
Comment (GMHDBJD) [inline loader/checkpoint.go:414]: terror.ErrLoadTaskChecPointNotMatch.AnnotateDelegate or terror.Annotate(err,...) We should keep the origin err.

Comment (Ehco1996) [reply loader/checkpoint.go:414]: - -. L414 is no origin error any more, should i use `errors.New("map key not found")` ?

**Review by GMHDBJD — APPROVED** [review]
LGTM

**Pull request merged**